stdlib/Scratch/src/Scratch.jl

```diff
@@ -0,0 +1,240 @@
+to that parent package, allowing the space to be garbage collected if all versions of the
+package that used it have been removed.
+
+If `parent_pkg` is not defined, or is a `Module` without a root UUID (e.g. `Main`,
+`Base`, an anonymous module, etc...) the created scratch space is parented to the global
+environment for the current version of Julia.
+
+Scratch spaces are removed if all parent projects that have accessed them are removed.
+As an example, if a scratch space is used by two versions of the same package but not a
+newer version, when the two older versions are removed the scratch space may be garbage
+collected.  See `Pkg.gc()` and `track_scratch_access()` for more details.
+"""
+function get_scratch!(key::AbstractString, parent_pkg::Union{UUID,Nothing} = nothing)
+    # Calculate the path and create the containing folder
+    path = scratch_path(key, parent_pkg)
+    mkpath(path)
+
+    # We need to keep track of who is using which spaces, so we track usage in a log
+    track_scratch_access(parent_pkg, path)
+    return path
+end
+function get_scratch!(key::AbstractString, parent_pkg::Module)
+    return get_scratch!(key, Base.PkgId(parent_pkg).uuid)
+end
+
+"""
+    delete_scratch!(key, parent_pkg)
+
+Explicitly deletes a scratch space created through `get_scratch!()`.
+"""
+function delete_scratch!(key::AbstractString, parent_pkg::Union{UUID,Nothing} = nothing)
+    path = scratch_path(key, parent_pkg)
+    rm(path; force=true, recursive=true)
+    delete!(scratch_access_timers, path)
+    return nothing
+end
+function delete_scratch!(key::AbstractString, parent_pkg::Module)
+    return delete_scratch!(key, Base.PkgId(parent_pkg).uuid)
+end
+
+"""
+    clear_scratchspaces!()
+
+Delete all scratch spaces in the current depot.
+"""
+function clear_scratchspaces!()
+    rm(scratch_dir(); force=true, recursive=true)
+    empty!(scratch_access_timers)
+    return nothing
+end
+
+"""
+    clear_scratchspaces!(parent_pkg::UUID)
+
+Delete all scratch spaces for the given package
+"""
+function clear_scratchspaces!(parent_pkg::UUID)
+    parent_prefix = scratch_dir(string(parent_pkg))
+    for path in collect(keys(scratch_access_timers))
+        if startswith(path, parent_prefix)
+            delete_scratch!(path)
+        end
+    end
+    # Next, clean up any other scratch dirs (we don't have to worry about resetting timers here)
+    rm(scratch_dir(string(parent_pkg)); force=true, recursive=true)
+    return nothing
+end
+
+"""
+    clear_scratchspaces!(m::Module)
+
+Delete all scratch spaces for the package that owns the given `Module`.  Throws an
+`ArgumentError` if the `Module` does not belong to a package.
+"""
+function clear_scratchspaces!(m::Module)
+    uuid = Base.PkgId(__module__).uuid
+    if uuid === nothing
+        throw(ArgumentError("Cannot find owning package for Module $(m)"))
+    end
+    return clear_scratchspaces!(uuid)
+end
+
+"""
+    @get_scratch!(key)
```

Comment:
Sorry for the drive by commenting: I really think it would be better not to have this macro in the API. As a package developer you wouldn't call `@get_scratch!` very often, so having to memorize a name once as a function and once as a macro is cumbersome (e.g. there has even been confusion about the often used `error()` and `@error`). At least it is harder to remember than the omnipresent `@__MODULE__`.

Reply:
I disagree; I think making it easy for the user to not have to look up their own UUID or pass in `@__MODULE__` all the time is worthwhile.